cpp/src/arrow/type.h

```diff
@@ -191,16 +192,33 @@ class ARROW_EXPORT Fingerprintable {
 
 }  // namespace detail
 
+/// EXPERIMENTAL: Layout specification for a data type
 struct ARROW_EXPORT DataTypeLayout {
-  // The bit width for each buffer in this DataType's representation
-  // (kVariableSizeBuffer if the item size for a given buffer is unknown or variable,
-  //  kAlwaysNullBuffer if the buffer is always null).
-  // Child types are not included, they should be inspected separately.
-  std::vector<int64_t> bit_widths;
-  bool has_dictionary;
-
-  static constexpr int64_t kAlwaysNullBuffer = 0;
-  static constexpr int64_t kVariableSizeBuffer = -1;
+  enum BufferKind { FIXED_WIDTH, VARIABLE_WIDTH, BITMAP, ALWAYS_NULL };
+
+  /// Layout specification for a single data type buffer
+  struct BufferSpec {
```

Comment:
Should we take the opportunity to encode the expected minimum size of the buffer as a function of the number of values (when possible)? This would be useful for, say, the offsets which have the `length+1` trick.

Reply:
Or there could be an additional `OFFSETS` BufferKind?

Reply:
You still can compute the size of bitmaps and fixed width buffers.

Reply:
`OFFSETS` BufferKind is also a good thing.